File: solutions/removeDuplicatesFromSortedArray-26/removeDuplicatesFromSortedArray-26.cpp

/**********************************************************************************
*
* Given a sorted array nums, remove the duplicates in-place such that each element appears only once and returns the new length.
*
* Input: nums = [1,1,2]
* Output: 2, nums = [1,2]
*
*
**********************************************************************************/

class Solution {

  /*
  *   [Two Pointers]
  *
  *   Time complexity : O(n). Assume that n is the length of array. Each of i and j traverses at most n steps.
  *
  *   Space Complexity: O(1).
  */

public:
    int removeDuplicates(vector<int>& nums) {
        
        int n = nums.size();
        if (n == 0) {
            return 0;
        }
        
        int left = 0;
        int right = 0;
        
        while(right < n){
            
            if (nums[right] != nums[left]) {
                left++;
                nums[left] = nums[right];
            }
            right++;
        }
        return left+1;
    }
};

File: solutions/removeDuplicatesFromSortedArray-26/removeDuplicatesFromSortedArray-26.py
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        
        n = len(nums)
        if n==0:
            return 0
        left = 0
        right = 0
        
        while right<n:
            
            if nums[right] != nums[left]:
                left += 1
                nums[left] = nums[right]
            right += 1
        
        return left+1